packages/perennial/contracts/multiinvoker/MultiinvokerRollup.sol

```diff
@@ -0,0 +1,290 @@
+                address product; uint256[] memory programIds;
+                (product, programIds, ptr) = decodeProductPrograms(input, ptr);
+
+                controller.incentivizer().claimFor(msg.sender, IProduct(product), programIds);
+            } else if (action == 8) { // WRAP 
+                address receiver; UFixed18 amount;
+                (receiver, amount, ptr) = decodeAddressAmount(input, ptr);
+
+                wrap(receiver, amount);
+            } else if (action == 9) { // UNWRAP
+                address receiver; UFixed18 amount;
+                (receiver, amount, ptr) = decodeAddressAmount(input, ptr);
+
+                unwrap(receiver, amount);
+            } else if (action == 10) { // WRAP_AND_DEPOSIT
+                address account; address product; UFixed18 amount;
+                (account, product, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+                
+               wrapAndDeposit(account, IProduct(product), amount);
+            } else if (action == 11) { // WITHDRAW_AND_UNWRAP
+                address receiver; address product; UFixed18 amount;
+                (receiver, product, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                withdrawAndUnwrap(receiver, IProduct(product), amount);
+            } else if (action == 12) { // VAULT_DEPOSIT
+                address depositer; address vault; UFixed18 amount;
+                (depositer, vault, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                vaultDeposit(depositer, IPerennialVault(vault), amount);
+            } else if (action == 13) { // VAULT_REDEEM
+                address vault; UFixed18 shares;
+                (vault, shares, ptr) = decodeAddressAmount(input, ptr);
+
+                IPerennialVault(vault).redeem(shares, msg.sender);
+            } else if (action == 14) { // VAULT_CLAIM
+                address owner; address vault;
+                (owner, vault, ptr) = decodeAddressAddress(input, ptr);
+
+                IPerennialVault(vault).claim(owner);
+            } else if (action == 15) { // VAULT_WRAP_AND_DEPOSIT 
+                address account; address vault; UFixed18 amount;
+                (account, vault, amount, ptr) = decodeAddressAddressAmount(input, ptr);
+
+                vaultWrapAndDeposit(account, IPerennialVault(vault), amount);
+            }
+        }
+    }
+
+    /// Example Calldata Structure
+    /// let ptr
+    /// [ptr(userLen), ptr+1:userLen(user registry # OR 20 byte address if userLen == 0)] => address user 
+    /// ptr += (userLen OR 20) + 1
+    /// [ptr(prodcutLen), ptr+1:productLen(product registry # OR 20 byte address if prdoctLen == 0)] => address product
+    /// ptr += (prodcutLen OR 20) + 1
+    /// [ptr(amountLen), ptr:amountLen] => uint256 amount 
+
+
+    // ARGUMENT TYPE DECODING //
+
+    function decodeAddressAddressAmount(bytes calldata input, uint ptr) private returns (address addr1, address addr2, UFixed18 amount, uint) {
```

Comment:
use `_` prefix on all `internal / private` functions

Reply:
see https://github.com/equilibria-xyz/perennial-mono/pull/141#discussion_r1133159867 re `wrap()` and `unwrap()`